### PR TITLE
[CDU] Init Page Revisions

### LINKED ISSUE #823: [META] MCDU/FMGS
* Will be updated with more specific tasks *

**Mod Version**

Master (separate branch will be created soon)

**Pages:**
- [ ] Data page
- [ ] Init A page
- [ ] Init B page
- [ ] Flight plan page
- [ ] Sec f-plan page
- [ ] Lat Rev page
- [ ] Vert Rev page
- [ ] Rad nav page
- [ ] Fuel pred page
- [ ] Prog page
- [ ] Perf page

**To Reproduce**

N/A

**Expected behavior**

N/A

**Actual behavior**

N/A

**References**

TODO (contact Ares#3394 for specific info/references if needed)

Discord username (if different from GitHub): Iceman#5540

### PR DESCRIPTION
**Pull request by Lucky38i**
#823
Progression to complete #689

**Summary of Changes**
- Colors changed to represent real-world aircraft across the entire MCDU
- Initial states for INIT A and INIT B revamped
- Introduced split colouring for INIT B (thanks Externo)
- Introduced small fonts on particular areas for INIT B
- Prevented auto computation of ZFW and ZFWCG
- Created parser for entering Trip wind, ZFW and ZFWCG, so you can enter in the various ways supported

If QA Testers could check this and let me know what is missing or needs to be changed

**Things left to do**
- Finish fuel computations #689 so INIT B and FUEL PRED are properly populated
- Add fuel planning in INIT B
- Per the Honeywell manual, certain sections will use big font if the user decides to overwrite computed values, this is not yet simulated in certain sections, will be done in #689

**Known Issues**
- Several values show as 0.00, this is a placeholder values until #689 is completed
- Co routes aren't when a from/to is inserted, this is a seperate issue on importing co routes.
- WIND page currently doesn't work, maybe someone can take this up otherwise, I might think of doing it after completing #689

**Screenshots (if necessary)**
image
image
image
image
## ZFW and Block has 2 blocks this has been fixed see 87d55a7
image 
image
image

**Additional context**
Honeywell FMC Manual
A320DP

Discord username: Lucky38#3550

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Ares1378** [comment]
Definitions:
LSK - LINE SELECT KEY on the MCDU
EXAMPLE : LSK 1L = LINE SELECT KEY 1 LEFT
________________________________________________________

There following items need to be corrected:
INIT A nothing entered real MCDU:
image
Changes that need to be made:
-INIT REQUEST:  this should be displayed next to LSK R2
-IRS INIT>: this should be displayed next to LSK R3
-LAT and LON data needs to be deleted this is never displayed on the MCDU INIT A page.
-LSK R6 Font is too big under TROPO

INIT A after FLT NBR and FROM/TO entered:
image
-CO RTE should be blank and not display NONE
-LAT and LON needs to be deleted 
-INIT REQUEST:  disappears from LSK R2
- IRS INIT> still displayed at LSK R3 

INIT A after being fully initialized:
image

INIT B nothing entered:
-trip wind should display HD000
image


INIT B ONCE ZFW/ZFWCG entered
-FUEL PLANNING --> in amber should appear next to LSK R3
image

INIT B when block fuel is entered:
-FUEL PLANNING --> disappears from LSK R3
image

INIT B after ZFW/ZFWCG and block fuel entered should look like the following:
image

Please let me know if you need anymore info.

Thanks,

Ares

**Comment by Benjozork** [comment]
Hello, please take note that we have started including a CHANGELOG.md file that requires every PR to add a change-log line that describes its changes. As your PR is not merged yet, we kindly ask that you add this information following the included directives in the file.

Disclaimer: This is a copy-and-pasted message.

**Comment by Lucky38i** [comment]
> 
> 
Hi Ares please see 39450a9 and below picture which shows the amended pages as you suggested
> There following items need to be corrected:
> INIT A nothing entered real MCDU:
> -LAT and LON data needs to be deleted this is never displayed on the MCDU INIT A page.
> -LSK R6 Font is too big under TROPO

image

> Changes that need to be made:
> -INIT REQUEST: this should be displayed next to LSK R2
> -IRS INIT>: this should be displayed next to LSK R3

This is being done here #677

> 
> WHEN FROM/TO ENTERED the page should auto appear and look like the following:

image

> INIT A after FLT NBR and FROM/TO entered:
> 
> -CO RTE should be blank and not display NONE
> -LAT and LON needs to be deleted

image

> -INIT REQUEST: disappears from LSK R2
> 
>     * IRS INIT> still displayed at LSK R3

This is being done here #677

> INIT A after being fully initialized:

image
 
> INIT B nothing entered:
> -trip wind should display HD000

image

> INIT B ONCE ZFW/ZFWCG entered
> -FUEL PLANNING --> in amber should appear next to LSK R3

image
 
> INIT B when block fuel is entered:
> -FUEL PLANNING --> disappears from LSK R3
> INIT B after ZFW/ZFWCG and block fuel entered should look like the following:

image


Thanks for the corrections Ares, really appreciate it. If there's anything else astray please let me know!

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js [hunk 1]
```diff
@@ -86,6 +94,108 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
...
+            this._windDir = "TL"
+            if (isFinite(wind)) return this.trySetAverageWind(wind)
+            else {
+                this.showErrorMessage("FORMAT ERROR")
+                return false
+            }
+        }
+    }
+
+    tryParseZeroFuelWeightZFWCG(s) {
```
Comment (lousybyte) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:167]: You can modify the `trySetZeroFuelWeightZFWCG` function directly if you include the needed file, https://github.com/flybywiresim/a32nx/pull/871/commits/a83295cabfec06054a220166378a8377f324a0f3 is exactly that, you can copy that commit to this PR.
Same goes for the custom `trySetAverageWind()` as well.

If you need to modify something that's in a shared file, follow the procedure from that commit to add the file to the mod.

**Review by lousybyte — CHANGES_REQUESTED** [review]
Also remove `layout.json` and `manifest.json` please.

**Review by lousybyte — APPROVED** [review]

**Comment by Lucky38i** [comment]
@AdenFlorian checked and tested

**Comment by AdenFlorian** [comment]
Please dont format entire files, it looks like you ran a formatter on FMCMainDispaly so now its heard to see what you changed. Please redo your changes to that file so that you only touch the lines that need to be touched for your changes, and double check that you started with the 1.9.3 version of that file, thanks!

**Comment by Lucky38i** [comment]
> Please dont format entire files, it looks like you ran a formatter on FMCMainDispaly so now its heard to see what you changed. Please redo your changes to that file so that you only touch the lines that need to be touched for your changes, and double check that you started with the 1.9.3 version of that file, thanks!

Fixed here ef5cb2d, as for FMCMainDisplay, I've grabbed the 1.9.3 changes and pasted it in the the file then added my changes. The file doesn't exist in master hence why it's hard to see the actual changes I made. Take a look here https://github.com/flybywiresim/a32nx/blob/ef5cb2d6880c6a5f69752d475cd7f33c789a1c92/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js#L1363-L1426

and here 
https://github.com/flybywiresim/a32nx/blob/ef5cb2d6880c6a5f69752d475cd7f33c789a1c92/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js#L1250-L1289

for the changes I made

**Comment by camilo625** [comment]
Name: Camilo625 #5575
Date of testing: Current date 02/10/2020
Version of the sim: 1.9.3.0
PR Tested: #888 

Testing technique: Sawn C&D and check Init Page
Steps you've made to test the Pull Request:
1. Spawn C%B
2. Load a FPL
3. Check everything its working

Overall Rating: Good
Conclusions of the testing: Seems to work fine, probably an IRL pilot should look at it to verify the correct behavior, other than that it looks good.

**Comment by lousybyte** [comment]
Inserting a value in the TAXI field seems to not update the displayed value, but the TOW is updated. Other than that, looks good!

**Comment by Pheome** [comment]
Tested with a short LHBP - LOWW route,C&D, simbrief generated FP added manualy.Everything worked fine.
Taxi is not updated as @lousybyte said,I don't know it's intended or not.Anyway it looks realy nice.

**Comment by Lucky38i** [comment]
@lousybyte @Pheome taxi issue has been resolved, thanks for spotting that.

**Comment by lousybyte** [comment]
Seems if the `ZFW/ZFWCG` is empty and you try to input just one of them, you get a `FORMAT ERROR`, not sure if that's intended or not, iirc is working fine w.o the PR.

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -21,5 +21,8 @@
 7. [PFD] Remove code related to unrelevant aircraft - @1Revenger1 (Avery Black)
 8. [ECAM] Lower ECAM DOOR Page Colour Fix - @nathaninnes (Nathan Innes)
 9. [ND] Add DME distances, VOR/ADF needles and functioning ADF2 - @blitzcaster (bltzcstr)
+1. [OVHD] Fixed Battery Indicator Colour - @nathaninnes (Nathan Innes)
 10. [OVHD] Fixed Battery Indicator Colour - @nathaninnes (Nathan Innes)
 11. [MISC] Removed Fuel Patch from MSFS Update 1.8.3 - @nathaninnes (Nathan Innes)
+1. [CDU] Revised INIT A Page - @lucky38i (Lucky38i)
+1. [CDU] Revised INIT B Page - @lucky38i (lucky38i)
```
Comment (NathanInnes) [inline .github/CHANGELOG.md:28]: Remove dupe & change to current month

**Comment by Lucky38i** [comment]
> Seems if the `ZFW/ZFWCG` is empty and you try to input just one of them, you get a `FORMAT ERROR`, not sure if that's intended or not, iirc `ZFW` only is working fine w.o the PR.

This is intended and confirmed by @Ares1378. zfw and zfwcg must first be entered before you can try entering either on its own

**Comment by lousybyte** [comment]
Please merge master again and correct the changelog to current month and will be good to merge!

**Comment by Ares1378** [comment]
FMS Pegasus Step 1A Pre-flight 4-23:
After initial entry, the zero fuel weight and zero fuel weight center of gravity are modified together or separately. When both values are entered, the values are separated by a slash ( / ).When only the zero fuel weight center of gravity is modified, the entry is preceded by a slash. Once ZFW and ZFWCG are entered, the values cannot be manually cleared. When the FMS transitions to the done phase, the ZFW and ZFWCG values are cleared automatically.

**Comment by Lucky38i** [comment]
> 
> 
> Please merge master again and correct the changelog to current month and will be good to merge!

All done!

**Comment by Lucky38i** [comment]
will #648 be closed if this is pulled btw?

**Comment by NathanInnes** [comment]
> will #648 be closed if this is pulled btw?

Already is closed

**Review by wuperguy — APPROVED** [review]

**Comment by AdenFlorian** [comment]
please resolve the initpagejs merge conflicts

**Comment by Lucky38i** [comment]
> 
> 
> please resolve the initpagejs merge conflicts

fixed conflicts and tested integration with #677. Working as intended

**Comment by Noxleee** [comment]
Name: Noxlee #7727
Date of testing: Current date 06/10/2020
Version of the sim: 1.9.3.0
PR Tested: #888 

Testing technique: Loaded up ac at gate EDDF. testing the MCDU Pages:
Checks I made:
1. Colours: Same as on the original screenshots, but effecting the whole MCDU

2. Init A: 
          - Lat and Lon are not displayed, see attached screenshot Init A
          - Co RTE is empty after inserting Airports, see attached screenshot Init A

3. Init B: 
          - Looks the same as the original screenshots when empty
          - When filled with data looks like attached screenshot Init B
          - TOW/LW -> the / is at TOW but should be at LW -> TOW      /LW


Overall Rating: to be improved
Conclusions of the testing: All in all its working, but a few things are misaligned and missing.


Init A
grafik

Init B
grafik

**Comment by tareksabet** [comment]
To confirm did you follow the Honeywell MCDU model which it should look like that image image

**Comment by Noxleee** [comment]
Yes it's looking like the honeywell MCDU , the original Screenshots of the PR are just old.

**Comment by AdenFlorian** [comment]
LGTM!

**Pull request merged**